graphbolt/src/csc_sampling_graph.cc

```diff
@@ -227,16 +271,39 @@ c10::intrusive_ptr<SampledSubgraph> CSCSamplingGraph::SampleNeighbors(
       probs_or_mask = probs_or_mask.value().to(torch::kFloat32);
     }
   }
+
   if (layer) {
     const int64_t random_seed = RandomEngine::ThreadLocal()->RandInt(
         static_cast<int64_t>(0), std::numeric_limits<int64_t>::max());
     SamplerArgs<SamplerType::LABOR> args{indices_, random_seed, NumNodes()};
-    return SampleNeighborsImpl(
-        nodes, fanouts, replace, return_eids, probs_or_mask, args);
+    if (fanouts.size() == 1) {
+      return SampleNeighborsImpl(
+          nodes, return_eids,
+          GetPickFn(
+              fanouts, replace, indptr_.options(), type_per_edge_,
+              probs_or_mask, args));
+    } else {
+      return SampleNeighborsImpl(
+          nodes, return_eids,
+          GetPickFnByEtype(
+              fanouts, replace, indptr_.options(), type_per_edge_,
+              probs_or_mask, args));
+    }
   } else {
     SamplerArgs<SamplerType::NEIGHBOR> args;
-    return SampleNeighborsImpl(
-        nodes, fanouts, replace, return_eids, probs_or_mask, args);
+    if (fanouts.size() == 1) {
+      return SampleNeighborsImpl(
+          nodes, return_eids,
+          GetPickFn(
+              fanouts, replace, indptr_.options(), type_per_edge_,
+              probs_or_mask, args));
+    } else {
+      return SampleNeighborsImpl(
+          nodes, return_eids,
+          GetPickFnByEtype(
+              fanouts, replace, indptr_.options(), type_per_edge_,
+              probs_or_mask, args));
+    }
```

Comment:
Is there any good way to reduce the repeated code? The problem is that the type of `args` in the two branches are different. The return types of `GetPickFn` and `GetPickFnByEtype` are also different.

Reply:
It is good for now. However, if one more template parameter is going to be added in the future, having 2^3 options, then we can consider either using macros (`#define`) or the use of template metaprogramming tricks to loop over all options using recursion. An example is here: https://stackoverflow.com/questions/52286202/dynamic-dispatch-to-template-function-c. There may be better examples or ways to do it than this one.

However, it is not getting out of hand for now. We can consider reducing code size later if needed. The bad thing about using templates is the dispatch can become ugly. The good thing is the excellent performance (since everything happens at compile time) and that only the dispatch is ugly, the main code is almost identical :).

Reply:
However, we might want to consider the following solution also: if having one more if condition at runtime doesn't bother us when it comes to performance, which it shouldn't as we do the same for the replace or weight parameters, then it is possible modify `GetPickFn` to move the `if (ByEtype)` check to inside of the lambda to eliminate `GetPickFn`'s `ByEtype` template argument. Then, a single lambda will be returned and we don't have to dispatch with respect to `ByEtype` anymore as the returned lambda will do the if check at runtime. In this case, since the if condition always succeeds or fails, it is easy for the branch predictor to get it right every time so shouldn't have any noticeable runtime impact.